README.md
# Detect GPU

[![npm version](https://badge.fury.io/js/detect-gpu-js.svg)](https://badge.fury.io/js/detect-gpu-js)
[![gzip size](https://img.badgesize.io/https:/unpkg.com/detect-gpu-js/dist/detect-gpu.esm.js?compression=gzip)](https://unpkg.com/detect-gpu-js)
[![install size](https://packagephobia.now.sh/badge?p=detect-gpu-js)](https://packagephobia.now.sh/result?p=detect-gpu-js)

Detect and rate GPUs based on their 3D rendering benchmark score allowing the developer to provide sensible default settings for graphically intensive applications. Think of it like a user-agent detection for the GPU but more powerful.

This project is a maintained fork of [detect-gpu](https://github.com/pmndrs/detect-gpu) project rewritten in JavaScript.

## Demo

[Live demo](https://soft8soft.github.io/detect-gpu-js/)

## Installation

Make sure you have [Node.js](http://nodejs.org/) installed, then:

```sh
 $ npm install detect-gpu-js
```

By default we use the [UNPKG](https://unpkg.com) CDN to host the benchmark data. If you would like to serve the benchmark data yourself, download the required benchmarking data from [benchmarks.tar.gz](https://github.com/soft8soft/detect-gpu-js/raw/master/benchmarks.tar.gz) and serve it from a public directory.

Alternatively, there is the `detect-gpu.bundle.js` build that comes with benchmark data embedded. This however requires 300KB+ more data downloaded over the network.

## Usage

```js
import { getGPUTier } from 'detect-gpu';

(async () => {
    const gpuTier = await getGPUTier();
  
    // Example output:
    // {
    //     "tier": 1,
    //     "isMobile": false,
    //     "type": "BENCHMARK",
    //     "fps": 21,
    //     "gpu": "intel iris graphics 6100"
    // }
})();
```

`detect-gpu-js` uses rendering benchmark scores (framerate, normalized by resolution) in order to determine what tier should be assigned to the user's GPU. If no `WebGLContext` can be created, the GPU is blocklisted or the GPU has reported to render on less than `15 fps` `tier: 0` is assigned. One should provide a fallback to a non-WebGL experience.

Based on the reported `fps` the GPU is then classified into either `tier: 1 (>= 15 fps)`, `tier: 2 (>= 30 fps)` or `tier: 3 (>= 60 fps)`. The higher the tier the more graphically intensive workload you can offer to the user.

## API

```js
getGPUTier({
    /**
     * URL of directory where benchmark data is hosted.
     *
     * @default https://unpkg.com/detect-gpu-js@{version}/dist/benchmarks
     */
    benchmarksURL,
    /**
     * Optionally pass in a WebGL context to avoid creating a temporary one
     * internally.
     */
    glContext,
    /**
     * Whether to fail if the system performance is low or if no hardware GPU is
     * available.
     *
     * @default false
     */
    failIfMajorPerformanceCaveat,
    /**
     * Framerate per tier for mobile devices.
     *
     * @defaultValue [0, 15, 30, 60]
     */
    mobileTiers,
    /**
     * Framerate per tier for desktop devices.
     *
     * @defaultValue [0, 15, 30, 60]
     */
    desktopTiers,
    /**
     * Use HiDPI (Retina) resolution to find benchmark results.
     *
     * @defaultValue true
     */
    hidpiScreenSize,
    /**
     * Optionally override specific parameters. Used mainly for testing.
     */
    override: {
        renderer,
        /**
         * Override whether device is an iPad.
         */
        isIpad,
        /**
         * Override whether device is a mobile device.
         */
        isMobile,
        /**
         * Override device screen size.
         */
        screenSize: { width, height },
        /**
         * Override how benchmark data is loaded.
         */
        loadBenchmarks
    };
})
```

## Support

Requires JavaScript modules to be supported, e.g. works on Chrome 61+, Firefox 60+, Safari 11+ (most versions dated back to 2017).

## Changelog

[Changelog](CHANGELOG.md)

## Licence

The project is released under the [MIT license](https://raw.githubusercontent.com/soft8soft/detect-gpu-js/master/LICENSE).

`detect-gpu-js` uses both mobile and desktop benchmarking scores from [https://gfxbench.com](https://gfxbench.com).


CHANGELOG.md
# CHANGELOG

# 1.0.5

- Support builds with embeeded benchmark data.

# 1.0.4

- Dropped support for Internet Explorer 11, Safari before version 11.

# 1.0.0

- Ported from from TypeScript to JavaScript.


src/index.js
import { version } from '../package.json';
// optional, when ignored in rollup config, becomes {}
import benchmarkFiles from './internal/benchmarks.json';

import { BLOCKLISTED_GPUS } from './internal/blocklistedGPUS';
import { cleanRenderer } from './internal/cleanRenderer';
import { deobfuscateRenderer } from './internal/deobfuscateRenderer';
import { deviceInfo } from './internal/deviceInfo';
import { OutdatedBenchmarksError } from './internal/error';
import { getGPUVersion } from './internal/getGPUVersion';
import { getLevenshteinDistance, tokenizeForLevenshteinDistance } from './internal/getLevenshteinDistance';
import { getWebGLContext } from './internal/getWebGLContext';
import { isSSR } from './internal/ssr';
import { isDefined } from './internal/util';
import { debug } from './internal/debug';

export const getGPUTier = async ({
            mobileTiers = [0, 15, 30, 60],
            desktopTiers = [0, 15, 30, 60],
            override = {},
            glContext,
            failIfMajorPerformanceCaveat = false,
            benchmarksURL = `https://unpkg.com/detect-gpu-js@${version}/dist/benchmarks`,
            hidpiScreenSize = true,
        } = {}) => {

    const queryCache = {};

    // populate cache from bundle if there is somemething in it
    for (const benchmarkFile in benchmarkFiles)
        queryCache[benchmarkFile] = benchmarkFiles[benchmarkFile].slice(1); // strip version

    if (isSSR) {
        return {
            tier: 0,
            type: 'SSR',
        };
    }

    const {
        isIpad = !!deviceInfo?.isIpad,
        isMobile = !!deviceInfo?.isMobile,
        screenSize = window.screen,
        loadBenchmarks = async (file) => {

            const data = await fetch(`${benchmarksURL}/${file}`).then((response) =>
                response.json()
            );

            // Remove version tag and check version is supported
            const version = parseInt(data.shift().split('.')[0], 10);
            if (version < 1) {
                throw new OutdatedBenchmarksError(
                    'Detect GPU benchmark data is out of date. Please update to version 1x'
                );
            }
            return data;
        },
    } = override;

    let { renderer } = override;

    const getGpuType = (renderer) => {
        const types = isMobile ?
            ['adreno', 'apple', 'mali-t', 'mali', 'nvidia', 'powervr', 'samsung'] :
            ['intel', 'apple', 'amd', 'radeon', 'nvidia', 'geforce', 'adreno'];
        for (const type of types) {
            if (renderer.includes(type)) {
                return type;
            }
        }
    };

    async function queryBenchmarks(renderer) {
        const type = getGpuType(renderer);
        if (!type) {
            return;
        }

        debug('queryBenchmarks - found type:', type);

        const benchmarkFile = `${isMobile ? 'm' : 'd'}-${type}${ isIpad ? '-ipad' : '' }.json`;

        const benchmark = (queryCache[benchmarkFile] = queryCache[benchmarkFile] ?? loadBenchmarks(benchmarkFile));

        let benchmarks;
        try {
            benchmarks = await benchmark;
        } catch (error) {
            if (error instanceof OutdatedBenchmarksError) {
                throw error;
            }
            debug("queryBenchmarks - couldn't load benchmark:", error);
            return;
        }

        const version = getGPUVersion(renderer);

        let matched = benchmarks.filter(
            ([, modelVersion]) => modelVersion === version
        );

        debug(`found ${matched.length} matching entries using version '${version}':`, matched.map(([model]) => model));

        // If nothing matched, try comparing model names:
        if (!matched.length) {
            matched = benchmarks.filter(([model]) => model.includes(renderer));

            debug(`found ${matched.length} matching entries comparing model names`, { matched, });
        }

        const matchCount = matched.length;

        if (matchCount === 0) {
            return;
        }

        const tokenizedRenderer = tokenizeForLevenshteinDistance(renderer);

        let [gpu, , , , fpsesByPixelCount] = matchCount > 1 ?
            matched
            .map((match) => [
                match, getLevenshteinDistance(tokenizedRenderer, match[2]),
            ])
            .sort(([, a], [, b]) => a - b)[0][0] : matched[0];

        debug(`${renderer} matched closest to ${gpu} with the following screen sizes:`, JSON.stringify(fpsesByPixelCount));

        let minDistance = Number.MAX_VALUE;
        let closest;

        let pixelCount = screenSize.width * screenSize.height;

        if (hidpiScreenSize)
            pixelCount = pixelCount * window.devicePixelRatio * window.devicePixelRatio;

        for (const match of fpsesByPixelCount) {
            const [width, height] = match;
            const entryPixelCount = width * height;
            const distance = Math.abs(pixelCount - entryPixelCount);

            if (distance < minDistance) {
                minDistance = distance;
                closest = match;
            }
        }

        if (!closest) {
            return;
        }

        const [, , fps, device] = closest;

        return [minDistance, fps, gpu, device];
    }

    const toResult = (tier, type, gpu, fps, device) => ({
        device,
        fps,
        gpu,
        isMobile,
        tier,
        type,
    });

    let renderers;
    let rawRenderer = '';

    if (!renderer) {
        const gl = glContext || getWebGLContext(deviceInfo?.isSafari12, failIfMajorPerformanceCaveat);
        if (!gl) {
            return toResult(0, 'WEBGL_UNSUPPORTED');
        }

        // NOTE: fixes console warning in Firefox
        const debugRendererInfo = deviceInfo?.isFirefox ? null : gl.getExtension('WEBGL_debug_renderer_info');

        if (debugRendererInfo) {
            renderer = gl.getParameter(debugRendererInfo.UNMASKED_RENDERER_WEBGL);
        } else {
            renderer = gl.getParameter(gl.RENDERER);
        }

        if (!renderer) {
            return toResult(1, 'FALLBACK');
        }

        rawRenderer = renderer;
        renderer = cleanRenderer(renderer);
        renderers = deobfuscateRenderer(gl, renderer, isMobile);
    } else {
        renderer = cleanRenderer(renderer);
        renderers = [renderer];
    }

    const results = (await Promise.all(renderers.map(queryBenchmarks)))
        .filter(isDefined)
        .sort(([aDis = Number.MAX_VALUE, aFps], [bDis = Number.MAX_VALUE, bFps]) =>
            aDis === bDis ? aFps - bFps : aDis - bDis
        );

    if (!results.length) {
        const blocklistedModel = BLOCKLISTED_GPUS.find(
            (blocklistedModel) => renderer.includes(blocklistedModel)
        );
        return blocklistedModel ?
            toResult(0, 'BLOCKLISTED', blocklistedModel) :
            toResult(1, 'FALLBACK', `${renderer} (${rawRenderer})`);
    }

    const [, fps, model, device] = results[0];

    if (fps === -1) {
        return toResult(0, 'BLOCKLISTED', model, fps, device);
    }

    const tiers = isMobile ? mobileTiers : desktopTiers;
    let tier = 0;

    for (let i = 0; i < tiers.length; i++) {
        if (fps >= tiers[i]) {
            tier = i;
        }
    }

    return toResult(tier, 'BENCHMARK', model, fps, device);
};
